dandi/dandiarchive.py

```diff
@@ -370,35 +407,65 @@ class _dandi_url_parser:
         """
-        Parse url like and return server (address), asset_id and/or directory
+        Parse a Dandi Archive URL and return a `ParsedDandiURL` instance.
 
-        Example URLs (as of 20210428):
+        The accepted URL formats are as follows.  Text in [brackets] is
+        optional.  A ``server`` field is a base API, GUI, or redirector URL for
+        a registered Dandi Archive instance.  If an optional ``version`` field
+        is omitted from a URL, the given Dandiset's most recent published
+        version, if any, or else its draft version will be used.
 
-        - Dataset landing page metadata:
-          https://gui.dandiarchive.org/#/dandiset/000003
+        - :samp:`https://identifiers.org/DANDI:{dandiset-id}` when it redirects
+          to one of the other URL formats
 
-        Individual and multiple files:
+        - :samp:`DANDI:{dandiset-id}` — Abbreviation for the above
 
-        - dandi???
+        - Any ``https://dandiarchive.org/`` or
+          ``https://*dandiarchive-org.netflify.app/`` URL which redirects to
+          one of the other URL formats
 
-        Multiple selected files + folders -- we do not support ATM, then further
-        RFing would be due, probably making this into a generator or returning a
-        list of entries.
+        - :samp:`https://{server}[/api]/[#/]dandiset/{dandiset-id}[/{version}][/files[?location={path}]]`
+          — If ``path`` is not specified, the URL refers to a Dandiset and is
+          converted to a `DandisetURL`.  If ``path`` is specified and it ends
+          with a forward slash, the URL refers to an asset folder and is
+          converted to an `AssetFolderURL` instance; if it does not end with a
+          slash, it refers to a single asset and is converted to an
+          `AssetItemURL` instance.
 
-        "Features":
+        - :samp:`https://{server}[/api]/dandisets/{dandiset-id}[/versions[/{version}]]`
+          — Refers to a Dandiset and is converted to a `DandisetURL`
 
-        - uses some of `known_instances` to map some urls, e.g. from
-          gui.dandiarchive.org ones into girder.
+        - :samp:`https://{server}[/api]/assets/{asset-id}[/download]` — Refers
+          to a single asset by identifier and is converted to a
+          `BaseAssetIDURL`
 
-        :rtype: ParsedDandiURL
+        - :samp:`https://{server}[/api]/dandisets/{dandiset-id}/versions/{version}/assets/{asset-id}[/download]`
+          — Refers to a single asset and is converted to an `AssetIDURL`
+
+        - :samp:`https://{server}[/api]/dandisets/{dandiset-id}/versions/{version}/assets/?path={path}`
+          — Refers to all assets in the given Dandiset whose paths begin with
+          the prefix ``path``; converted to an `AssetPathPrefixURL`
+
+        - :samp:`dandi://{instance-name}/{dandiset-id}[@{version}][/{path}]`,
+          where ``instance name`` is the name of a registered Dandi Archive
```

Comment:
```suggestion
          where ``instance-name`` is the name of a registered Dandi Archive
```
?

but also wondered if it wouldn't be better to use python's `_` instead of `-` even though we might not have `dandiset_id` or `instance_name` variables?